### PR TITLE
Record Data State

### PR DESCRIPTION
**Pull request by igorT**
[Rendered](https://github.com/igort/rfcs/blob/igor/record-state/text/0463-record-data-state.md)

### CODE REVIEW

**Comment by mikkopaderes** [comment]
Are there any plans for an `isUnloaded` state? Current use-case is for an adapter that supports realtime changes (e.g. Firebase). Whenever someone loses permission or a record was deleted some adapters might want to unload those records in realtime. However, I don't think there's anything that exists to indicate this state.

Another related property may be `unloadReason` where it'll contain the reason why the record is in `isUnloaded: true` state.

**Comment by runspired** [comment]
@mikkopaderes the short answer is that there is not. We did explore whether `recordData` should consider `unload` as an operation on `RecordState`, and the clear answer was **`"no"`**.

A TL;DR of what to do instead is ______________

If we start with a bit of context on what `unloadRecord` is _meant_ to achieve and how that fits in both the pre-record-data and post-record-data worlds helps to understand ______________

In pre-record-data `ember-data`, `InternalModel` fulfilled many roles including:

* being the identifier for a resource that may or may not have been loaded
   Read the [Identifiers RFC](https://github.com/emberjs/rfcs/pull/403) for more context and insight on how we are replacing and improving that role.

* being the cache for attributes/relationships for a resource (now handled by `RecordData`)

* being the cache for loading & network state: see [Request State RFC](https://github.com/emberjs/rfcs/pull/466)

* being the cache for `RecordState` (this RFC)

* being the cache for `Error State` (but not errors themselves which were stashed on record instances)
  Read the [RecordData Errors RFC](https://github.com/emberjs/rfcs/pull/465)

An unfortunate design within `InternalModel` was that `loading|network|error|record` state were mixed into a single state machine. You can see (from the RFCs above) that this is something we are working to rectify.

With the `InternalModel` design, `unloadRecord` played a _poorly fit_ role of needing to change the `state` of the `InternalModel` from wherever it was back into `empty`, needing to notify any relate records that it would need to be loaded again if re-accessed, and of actually emptying the cache.

In the `RecordData` design, unloadedRecords are simply absent from the cache, with knowledge of their existence maintained if needed via an `Identifier`. By separating the identity from the data, we can seamlessly release the data, and use the absence of it to know we need to retrieve it when asked.

File: text/0000-record-data-state.md [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+    
+    interface RecordData {
+    
+    	// To be added in this RFC
+    	isNew(identifier: RecordIdentifier): boolean
+    
+    	isDeleted(identifier: RecordIdentifier): boolean
+    
+      isDeletionPersisted(identifier: RecordIdentifier): boolean
```
Comment (rwjblue) [inline text/0000-record-data-state.md]: Bizarre spacing here, maybe tabs vs spaces or something?

File: text/0000-record-data-state.md [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+  unloadRecord(identifier: RecordIdentifier);
+
+  rollbackAttributes(identifier: RecordIdentifier);
+  changedAttributes(identifier: RecordIdentifier): ChangedAttributesHash;
+  hasChangedAttributes(identifier: RecordIdentifier): boolean;
+
+  getAttr(identifier: RecordIdentifier, key: string): any;
+  getRelationship(identifier: RecordIdentifier): JsonApiRelationship;
+
+  performMutation(operation: Operation): void
```
Comment (igorT) [inline text/0000-record-data-state.md]: Remove from RFC before merging

**Pull request merged**